[LoongArch64] coreclr-vm directory

shushanhf opened the pull request:
Part6-2: Add the coreclr-vm directory for LoongArch64.

shushanhf committed:
Merge pull request #1 from dotnet/main

update from runtime.

shushanhf committed:
Merge branch 'dotnet:main' into master

shushanhf committed:
Merge branch 'dotnet:main' into master

shushanhf committed:
Merge branch 'dotnet:main' into master

shushanhf committed:
Merge branch 'dotnet:main' into master

shushanhf committed:
Merge branch 'dotnet:main' into master

shushanhf committed:
Merge branch 'dotnet:main' into master

shushanhf committed:
Part6-2: -add the coreclr-vm directory for LoongArch64. (#59561)

Co-authored-by: Loongson's .NET-teams

shushanhf committed:
Merge branch 'dotnet:main' into main

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=62885) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=62885) before we can accept your contribution.<br/><br/>:x: shushanhf [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=62885)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=62885) it.</sub>

jkotas commented on src/coreclr/vm/syncblk.cpp:
```diff
@@ -532,6 +534,10 @@ SyncBlockCache::~SyncBlockCache()
     }
 }
 
+#ifdef TARGET_LOONGARCH64
```
Is it possible to use the LOONGARCH64 version everywhere without ifdef?

shushanhf replied:
> Is it possible to use the LOONGARCH64 version everywhere without ifdef?

OK, here is just for optimize.
If need, I can rewrite it.

Commit:
[LoongArch64] revert the syncblk.cpp.

kant2002 commented on src/coreclr/vm/ceeload.h:
```diff
@@ -87,6 +87,8 @@ class JITInlineTrackingMap;
 #elif defined(HOST_ARM64)
 #define NATIVE_SYMBOL_READER_DLL W("Microsoft.DiaSymReader.Native.arm64.dll")
+#elif defined(HOST_LOONGARCH64)
+#define NATIVE_SYMBOL_READER_DLL W("Microsoft.DiaSymReader.Native.loongarch64.dll")
```
Wow. Does this part of standard Nuget package, or this is something custom?

shushanhf replied:
And frankly, here I just add LoongArch64's supporting while don't understand it's usage.

kant2002 replied:
This DLL is Windows only as far as I aware. Not sure why this change is even needed. Better wait if this PR can be make green based on other comments in sibling PRs

shushanhf replied:
> This DLL is Windows only as far as I aware. Not sure why this change is even needed. Better wait if this PR can be make green based on other comments in sibling PRs

I will revert on this PR.
Thanks.

kant2002 commented on src/coreclr/vm/class.cpp:
```diff
@@ -1430,6 +1430,9 @@ MethodDesc* MethodTable::GetExistingUnboxedEntryPointMD(MethodDesc *pMD)
 bool MethodTable::IsHFA()
 {
     LIMITED_METHOD_CONTRACT;
+#if defined(TARGET_LOONGARCH64)
```
`FEATURE_HFA` seems to be not defined on LoongAarch64 so this whole block should not exists. If HFA not required, maybe better make sure that in `coreclr/inc/switches.h` `FEATURE_HFA` is not defined.

shushanhf replied:
> `FEATURE_HFA` seems to be not defined on LoongAarch64 so this whole block should not exists. If HFA not required, maybe better make sure that in `coreclr/inc/switches.h` `FEATURE_HFA` is not defined.

Yes, I will update it.
Thanks for your review~

kant2002 commented on src/coreclr/vm/comdelegate.cpp:
```diff
@@ -330,6 +339,89 @@ BOOL AddNextShuffleEntryToArray(ArgLocDesc sArgSrc, ArgLocDesc sArgDst, SArray<S
+
+                assert(!"---------No this stack case.--------------");
+            }
+            //else if (sArgSrc.m_cGenReg > 0)
```
Probably this section require some cleanup. Looks a bit in-development.

shushanhf replied:
> Probably this section require some cleanup. Looks a bit in-development.

Yes, Thanks for your review again.

kant2002 commented on src/coreclr/vm/interpreter.cpp:
```diff
@@ -91,7 +91,7 @@ InterpreterMethodInfo::InterpreterMethodInfo(CEEInfo* comp, CORINFO_METHOD_INFO*
 #endif
 
-#if defined(UNIX_AMD64_ABI)
+#if defined(UNIX_AMD64_ABI) || defined(HOST_LOONGARCH64)
```
Should this be `TARGET_LOONGARCH64`? Or maybe better if custom ABI is used, create `UNIX_LOONGARCH64_ABI`.

shushanhf replied:
For LoongArch64 there is only one ABI on linux, so I used the TARGET_LOONGARCH64 as the next define,
liking the Line:100  #elif defined(HOST_ARM) || defined(HOST_AMD64)|| defined(HOST_ARM64)

Is this OK?

am11 replied:
I think it is ok. You can also just add `UNIX_LOONGARCH64_ABI` to the target and bring it to the same plan as other architectures: https://github.com/dotnet/runtime/blob/0206926cf3e7d45804c66cbd282de094f69ecf61/src/coreclr/clrdefinitions.cmake#L233
(it will not be wrong since there is no ambiguity)

Note, this interpreter is experimental and it just compiles, it is not used anywhere and its tests were disabled in the CI few months ago. So if interpreter.cpp is compiling on LA64 with fewer and less different tweaks, it is fine for now.

shushanhf replied:
Thanks, @kant2002 @am11 
I will do it and upload the patch later.

kant2002 commented on src/coreclr/vm/interpreter.cpp:
```diff
@@ -882,6 +898,10 @@ CorJitResult Interpreter::GenerateInterpreterStub(CEEInfo* comp,
         // x8 through x15 are scratch registers on ARM64.
         IntReg x8 = IntReg(8);
         IntReg x9 = IntReg(9);
+#elif defined(HOST_LOONGARCH64)
```
Does this section needed?

shushanhf replied:
> Does this section needed?

I argree with you.
Maybe the other CPUs are the same case.

kant2002 reviewed (commented):
Left some comments.

shushanhf commented on src/coreclr/vm/interpreter.cpp:
```diff
@@ -882,6 +898,10 @@ CorJitResult Interpreter::GenerateInterpreterStub(CEEInfo* comp,
         // x8 through x15 are scratch registers on ARM64.
         IntReg x8 = IntReg(8);
         IntReg x9 = IntReg(9);
```
All these are local vaues and are unused codes.
So the LoongArch64 is just a noting and is ok to be deleted.

Commit:
[LoongArch64] delete some unused codes.

Commit:
[LoongArch64] add vm/CMakeLists.txt from #62889.

jkotas commented on src/coreclr/vm/callhelpers.cpp:
```diff
@@ -471,7 +471,11 @@ void MethodDescCallSite::CallTargetWorker(const ARG_SLOT *pArguments, ARG_SLOT *
                     case 1:
                     case 2:
                     case 4:
+#if defined(TARGET_LOONGARCH64)
```
If this is trying to extend the value to 64-bit, I do not think it is sufficient. The value should be sign or zero extended depending on whether the argument is signed or unsigned.

shushanhf replied:
> If this is trying to extend the value to 64-bit, I do not think it is sufficient. The value should be sign or zero extended depending on whether the argument is signed or unsigned.

Maybe this is a another question for different types CPU-architecture.
For AMD64 and ARM64 which they are 64-bits processors, the operation of the 32-bits values usually does not care the hight 32-bits.
But for the LoongArch64,MIPS64 (maybe the RISC-V-64),  the operation of the 32-bits values usually must care the hight 32-bits.

By the way, now the CoreCLR seems uncovered the non-AMD64/ARM64 type architectures.
Liking the comparation instruction, the operation's restriction for common data types.

jkotas replied:
Yes, the ABIs for existing CoreCLR architectures has unspecified value for the padding bits when primitive types are passed in registers. LoongArch is the first architecture that requires predictable content of the padding bits. This will likely require fixes in number of places.

In places like this one that are not performance sensitive, it should be fine to initialize the padding bits unconditionally (without ifdef). It will make the code simpler.

Also, I expect that this will be a source of subtle bugs for DllImport signatures. It is not unusual to see DllImport signature with signed/unsigned mismatch in .NET code out there (e.g. .NET side is defined as signed `int` and unmanaged C side is defined as `unsigned int`). DllImport with signed/unsigned mismatch works fine on all existing CoreCLR architectures today, but it won't be the case for LoongArch. Did you happen to run into any DllImport bugs like this during the port?

shushanhf replied:
> Yes, the ABIs for existing CoreCLR architectures has unspecified value for the padding bits when primitive types are passed in registers. LoongArch is the first architecture that requires predictable content of the padding bits. This will likely require fixes in number of places.
> 
> In places like this one that are not performance sensitive, it should be fine to initialize the padding bits unconditionally (without ifdef). It will make the code simpler.

good idea.
Thanks.

> 
> Also, I expect that this will be a source of subtle bugs for DllImport signatures. It is not unusual to see DllImport signature with signed/unsigned mismatch in .NET code out there (e.g. .NET side is defined as signed `int` and unmanaged C side is defined as `unsigned int`). DllImport with signed/unsigned mismatch works fine on all existing CoreCLR architectures today, but it won't be the case for LoongArch. Did you happen to run into any DllImport bugs like this during the port?

Yes, you are right.
If you had write a project, e.g. implenmened by C/C++, while you just considered the x64/arm64 and didn't care about 64the sign/unsign int type,  this may be ok on x64/arm64. But maybe failded on LoongArch64/MIPS64/RISC-V-64.
We had tested the CoreCLR calling the so libraries on linux, if the C# and the C/C++ are same defination, the arguments are the same with sigin-int or unsign-int at the same time, that will be ok.

jkotas commented on src/coreclr/vm/argdestination.h:
```diff
@@ -83,6 +83,37 @@ class ArgDestination
+    void CopyStructToRegisters(void *src, int fieldBytes)
+    {
+        _ASSERTE(IsStructPassedInRegs());
+        _ASSERTE(m_argLocDescForStructInRegs->m_cFloatReg == 1);
```
Where is the case of two floating registers handled?

shushanhf replied:
> Where is the case of struct passed in two floating registers handled?

A struct containing just one floating-point real is passed as though it were a standalone floating-point real.

This case is the same with the RISC-V: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc

kant2002 commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -9384,6 +9393,422 @@ CorInfoTypeWithMod CEEInfo::getArgType (
 }
 
+//Only used for LoongArch64-ABI.
+uint32_t CEEInfo::getFieldTypeByHnd(CORINFO_CLASS_HANDLE cls)
```
I believe this should be accompanied by `ThunkInput.txt` related changes from https://github.com/dotnet/runtime/pull/62893 That way you can make code compile in this PR. And all changes to this file (`src/coreclr/vm/jitinterface.cpp`) from same PR can be placed here. At least in this change it would be easier to justify JIT interface changes, rationale would be in the code usage.

shushanhf replied:
> I believe this should be accompanied by `ThunkInput.txt` related changes from #62893 That way you can make code compile in this PR. And all changes to this file (`src/coreclr/vm/jitinterface.cpp`) from same PR can be placed here. At least in this change it would be easier to justify JIT interface changes, rationale would be in the code usage.

Thanks,
You are right. I have to update all related files together.

shushanhf replied:
Maybe compiling sucessfully by CI is hard, Becasuse there are some related files committed by other PRs.
Especially the ABI's files are scattered.

kant2002 replied:
I hear you. I see how ABI related changes, and JIT interface are used here and that. I think it's important to at least make sure that there are no compilation issues at this stage. Do you have any ideas how this (and sibling) code can be sliced in better way for review?

shushanhf replied:
> I hear you. I see how ABI related changes, and JIT interface are used here and that. I think it's important to at least make sure that there are no compilation issues at this stage. Do you have any ideas how this (and sibling) code can be sliced in better way for review?

I think adding a new CPU-architecture is different with committing a small patch for PR.
Maybe there is a special way for how to commit and review for adding a new architecture.

Liking the compilation issues, maybe compiling all the patches together is a good way at least for adding a new CPU-architecuture.

But for review the patches, it's better that committing several patches.
That‘s a contradiction.

jkotas replied:
I think that the challenge here is due to large code change that was developed offline for a long period of time. Large deltas like that are always challenging to merge upstream.

This is not the first new architecture port that we got for .NET runtime. The difference is that the previous ports (e.g. Linux x86 or ARM64) were not developed offline. The development for the previous ports was in the upstream repo, that naturally broke them into small manageable PRs.

I would recommend to focus on making forward progress. Break the work into smaller PRs as necessary, get those smaller PRs passing CI, reviewed and merged. Rinse and repeat until the whole port is in.

shushanhf replied:
> I think that the challenge here is due to large code change that was developed offline for a long period of time. Large deltas like that are always challenging to merge upstream.
> 
> This is not the first new architecture port that we got for .NET runtime. The difference is that the previous ports (e.g. Linux x86 or ARM64) were not developed offline. The development for the previous ports was in the upstream repo, that naturally broke them into small manageable PRs.
> 
> I would recommend to focus on making forward progress. Break the work into smaller PRs as necessary, get those smaller PRs passing CI, reviewed and merged. Rinse and repeat until the whole port is in.

Yes, now we had done it.
For this PR, I have to ensure this PR compiling ok first, is it right ?

shushanhf replied:
Now based on the latest patch, this PR can be compiled sucessfully.
And I will fix all the compiling error within every PRs.

kant2002 reviewed (commented):
Some comments about compilation issues.

shushanhf commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -654,7 +669,7 @@ UINT_PTR ExceptionTracker::CallCatchHandler(CONTEXT* pContextRecord, bool* pfAbo
     {
         _ASSERTE(m_uCatchToCallPC != 0 && m_pClauseForCatchToken != NULL);
-        uResumePC = CallHandler(m_uCatchToCallPC, sfStackFp, &m_ClauseForCatch, pMD, Catch X86_ARG(pContextRecord) ARM_ARG(pContextRecord) ARM64_ARG(pContextRecord));
+        uResumePC = CallHandler(m_uCatchToCallPC, sfStackFp, &m_ClauseForCatch, pMD, Catch X86_ARG(pContextRecord) ARM_ARG(pContextRecord) ARM64_ARG(pContextRecord) LOONGARCH64_ARG(pContextRecord));
```
> If add `src/coreclr/inc/stdmacros.h` from #62886 then this part would compile

Thanks, I will updae later.

Commit:
[LoongArch64] add related files from #62886 and #62893.

Commit:
[LoongArch64] moved vm/jitinterface.cpp from #62893.
moved inc/corinfo.h from #6288.

shushanhf committed:
[LoongArch64] run the file `src/coreclr/tools/Common/JitInterface/ThunkGenerator/gen.sh`
and merge the patch from #62893.

shushanhf committed:
[LoongArch64] revert vm/ceeload.h.
amend the `ToolBox/superpmi/superpmi-shared/agnostic.h`

shushanhf commented:
Now this PR can be compiled sucessfully on the native mode.
https://dev.azure.com/dnceng/public/_build/results?buildId=1520952&view=logs&j=0e64859c-a870-5a74-6b6e-333fc1003298&t=3d90743c-2c9e-5ca7-807c-f73e0e661864

What should I do next step ?

kant2002 commented:
> Now this PR can be compiled sucessfully on the native mode. 

Great news

> What should I do next step ?

I think you should add here https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs something like that
```csharp
private CorInfoTypeWithMod getArgType2(CORINFO_SIG_INFO* sig, CORINFO_ARG_LIST_STRUCT_* args, CORINFO_CLASS_STRUCT_** vcTypeRet, ref int flags)
{
    throw new NotImplementedException("For LoongArch64, would be implemented later");
}

private uint getFieldTypeByHnd(CORINFO_CLASS_STRUCT_* cls)
{
    throw new NotImplementedException("For LoongArch64, would be implemented later");
}
```
that should move you forward with fixing compilation errors in managed code.

We should resolve all compilation issues here before any changes can land main. Otherwise at least dozen of people cannot do work.

I think there good chance that CI become green if `./build.sh` also did not produce any errors locally. That's how I test my suggestions to you.

shushanhf commented:
> > Now this PR can be compiled sucessfully on the native mode.
> 
> Great news
> 
> > What should I do next step ?
> 
> I think you should add here https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs something like that
> 
> ```cs
> private CorInfoTypeWithMod getArgType2(CORINFO_SIG_INFO* sig, CORINFO_ARG_LIST_STRUCT_* args, CORINFO_CLASS_STRUCT_** vcTypeRet, ref int flags)
> {
>     throw new NotImplementedException("For LoongArch64, would be implemented later");
> }
> 
> private uint getFieldTypeByHnd(CORINFO_CLASS_STRUCT_* cls)
> {
>     throw new NotImplementedException("For LoongArch64, would be implemented later");
> }
> ```
> 
> that should move you forward with fixing compilation errors in managed code.
> 
> We should resolve all compilation issues here before any changes can land main. Otherwise at least dozen of people cannot do work.
> 
> I think there good chance that CI become green if `./build.sh` also did not produce any errors locally. That's how I test my suggestions to you.

Thanks, I will update it right now.

Commit:
[LoongArch64] add empty interfaces within `CorInfoImpl.cs`.

Commit:
[LoongArch64] Fix the compiling error on Windows.

kant2002 commented on src/coreclr/vm/comdelegate.cpp:
```diff
@@ -193,7 +202,7 @@ class ShuffleIterator
             const unsigned byteIndex = m_argLocDesc->m_byteStackIndex + m_currentByteStackIndex;
 
-#if !defined(TARGET_OSX) || !defined(TARGET_ARM64)
+#if !defined(TARGET_OSX) || !defined(TARGET_ARM64) || !defined(TARGET_LOONGARCH64)
```
@shushanhf I suspect that this line cause current OX issues on tests.

shushanhf replied:
Yes, there is only two failed on the CI.

shushanhf replied:
Assert failure(PID 44877 [0x0000af4d], Thread: 22157090 [0x1521722]): byteIndex == unsigned(index * TARGET_POINTER_SIZE)
    File: /Users/runner/work/1/s/src/coreclr/vm/comdelegate.cpp Line: 216
    Image: /private/tmp/helix/working/A8030972/p/corerun

kant2002 replied:
I believe you don't need this line for `LOONGARCH64` . It skip subsequent code on OSX with M1 processor. Do you need to skip that code on LoongArch64 too?

shushanhf replied:
Maybe  I find the reason.
I will update it and try again.

shushanhf replied:
@kant2002 
Thanks for your review and help.

I upload new patch.

kant2002 replied:
I think you should try to run these specific tests localy. Just to double-check before this is fails. Not sure if you try that already or not.
```
./src/tests/build.sh --tree JIT/SIMD
./src/tests/run.sh
```

that way you build just specific tests and run only tests in JIT/SIMD folder.

kant2002 replied:
> I believe you don't need this change for LOONGARCH64 . It skip subsequent code on OSX with M1 processor. Do you need to skip that code on LoongArch64 too?

If you don't need to skip this code for LoongArch64, can you undo change in this line, to restore back OSX M1 tests?

shushanhf replied:
> I believe you don't need this change for `LOONGARCH64` . It skip subsequent code on OSX with M1 processor. Do you need to skip that code on LoongArch64 too?

Yes, you are right.
I had updated it.
For LoongArch64, I should use this block.

Commit:
[LoongArch64] Fix the compiling error for memory unaligned m_currentByteStackIndex.

jkotas commented on src/coreclr/vm/comdelegate.cpp:
```diff
@@ -330,6 +339,73 @@ BOOL AddNextShuffleEntryToArray(ArgLocDesc sArgSrc, ArgLocDesc sArgDst, SArray<S
     {
         // We should have slots to shuffle in the destination at the same time as the source.
         _ASSERTE(iteratorDst.HasNextOfs());
+#if defined(TARGET_LOONGARCH64)
```
Can this logic be in `ShuffleIterator` iterator instead? All other platforms have the platform-specific logic in `ShuffleIterator`.

shushanhf replied:
> Can this logic be in `ShuffleIterator` iterator instead? All other platforms have the platform-specific logic in `ShuffleIterator`.

Yes, this is needed for LoongArch64's ABI.
Liking LoongArch64's ABI, we have to get the details info of the struct's fields, where the other architectures now, liking X64/arm64, are not need these.

jkotas replied:
Can the special casing of `LoongArch64` be done inside the ShuffleIterator ?

For example, Linux x64 ABI (`UNIX_AMD64_ABI`) is even more complex than LoongArch64 and these details are hidden inside the `ShuffleIterator`.

shushanhf replied:
I will think about it.

shushanhf replied:
@jkotas   Sorry, I forgot this comment.
Thanks, I will update it later.

shushanhf replied:
I will delete this as the logic condition of the two `if statement` are confict which is not used for LoongArch64 now.
```
        if (!sArgDst.m_byteStackSize && (sArgSrc.m_structFloatFlag > 0))
        {
            if (sArgSrc.m_byteStackSize > 0)
```

jkotas commented on src/coreclr/vm/reflectioninvocation.cpp:
```diff
@@ -1022,6 +1022,17 @@ FCIMPL5(Object*, RuntimeMethodHandle::InvokeMethod,
         InvokeUtil::CopyArg(th, &objs->GetAt(i), &argDest);
+
+#if defined(TARGET_LOONGARCH64)
+        if ((ofs < 0) && (objs->GetAt(i) != 0) && (th.GetVerifierCorElementType() == ELEMENT_TYPE_VALUETYPE) && (th.GetSize() == 8) && (argit.m_idxFPReg == 2))
```
What is the situation that this is trying to handle? It does not look right to have this logic here - it seems to be compensating for more fundamental bug somewhere else.

shushanhf replied:
This is needed for the type ABI liking LoongArch64.
Here it handled the argment of struct {float a; float b;} which using two float-registers passing the arguments.
We have to treat this struct as two arguments so split it.

jkotas replied:
I understand that these structs require special handling for LoongArch64.

This place does not look like the right place to add this special handling. It should be done by changing the implementation of `ArgDestination` and  `UnBoxIntoArg` instead.

Also, according to https://github.com/loongson/LoongArch-Documentation/pull/32#discussion_r773539198, I think the structures with two doubles need the same treatment.

shushanhf replied:
> Also, according to [loongson/LoongArch-Documentation#32 (comment)](https://github.com/loongson/LoongArch-Documentation/pull/32#discussion_r773539198), I think the structures with two doubles need the same treatment.

Double doesn't need because it is a pointer size.

> This place does not look like the right place to add this special handling. It should be done by changing the implementation of `ArgDestination` and `UnBoxIntoArg` instead.

I didn't do that because there are not enought info to do it within the ArgDestination.
For example, ArgDestination can't get the info of the struct.

jkotas replied:
You can modify `ArgDestination` and `ArgLocDesc` as necessary for `LoongArch64` to capture the necessary details. Similar to how these types are modified for Linux x64 ABI.

shushanhf replied:
> You can modify `ArgDestination` and `ArgLocDesc` as necessary for `LoongArch64` to capture the necessary details. Similar to how these types are modified for Linux x64 ABI.

I can not modify the `ArgDestination` for LoongArch64 only before you accepting to do it.
Now If I modify the `ArgDestination`, there will be some modification that related with other architecuturs.
Should I do these  within this PR ?

jkotas replied:
I think it would be a good idea to have a separate PR with the VM calling convention related changes: modify `ArgDestination` and `ArgLocDesc`, add code to sign-extend and zero-extended CallDescr worker arguments as necessary, etc.

shushanhf replied:
That means I create a new PR for modifing ArgDestination and ArgLocDesc,
this PR should wait for the modify of ArgDestination and ArgLocDesc` finished.

How should I commit new patch based on the modify of ArgDestination and ArgLocDesc` ?
Because this PR is based on a old main, how can I resolve the conflics.

shushanhf replied:
That means I create a new PR for modifing ArgDestination and ArgLocDesc,
this PR should wait for the modify of ArgDestination and ArgLocDesc` finished.

How should I commit new patch based on the modify of ArgDestination and ArgLocDesc` ?
Because this PR is based on a old main, how can I resolve the conflics.

shushanhf replied:
I had deleted the modification within the file `src/coreclr/vm/reflectioninvocation.cpp`
by the commit
https://github.com/dotnet/runtime/pull/62885/commits/7757e09a9c190813241e7ac02b3436e1e9001471

jkotas commented on src/coreclr/vm/callhelpers.cpp:
```diff
@@ -471,7 +471,11 @@ void MethodDescCallSite::CallTargetWorker(const ARG_SLOT *pArguments, ARG_SLOT *
                     case 2:
                     case 4:
+#if defined(TARGET_LOONGARCH64)
+                        *((INT64*)pDest) = (INT32)pArguments[arg];
```
This change is not right now that we have established in https://github.com/loongson/LoongArch-Documentation/pull/32 that signed integers should be sign-extended and unsigned integers should be zero-extended.

shushanhf replied:
https://github.com/loongson/LoongArch-Documentation/pull/32#discussion_r772267496

jkotas replied:
The comment you have quoted talks about one specific example. The specific example has signed integer and so value is sign-extended.

The general rule earlier in the current version of the ABI doc is:

When passed in registers or on the stack, the unsigned integer scalars narrower than GRLEN bits are zero-extended to GRLEN bits, and the signed integer scalars are sign-extended.

https://github.com/loongson/LoongArch-Documentation/pull/32/files#diff-711b3e7b6a005b492898ac6d93f2d8d37c00e0831e210993a6f9dbb26c043717R711

This rule has to be applied here and in other places.

shushanhf replied:
> The comment you have quoted talks about one specific example. The specific example has signed integer and so the value is sign-extended.
> 
> The general rule in the current version of the ABI doc is:
> 
> > When passed in registers or on the stack, the unsigned integer scalars narrower than GRLEN bits are zero-extended to GRLEN bits, and the signed integer scalars are sign-extended.
> 
> https://github.com/loongson/LoongArch-Documentation/pull/32/files#diff-711b3e7b6a005b492898ac6d93f2d8d37c00e0831e210993a6f9dbb26c043717R711
> 
> This rule has to be applied here and in other places.

Thanks, I agree with you.
It's better to fit the signed and unsigned.
I will amend it.

Commit:
[LoongArch64] Delete the !TARGET_LOONGARCH64 for m_currentByteStackIndex.

Commit:
[LoongArch64] move ToolBox from #62886.

shushanhf commented:
Hi, @jkotas 
What should I do next ?

Commit:
[LoongArch64] amend the args when needs unsigned extending within CallTargetWorker.

jkotas commented on src/coreclr/ToolBox/superpmi/mcs/verbildump.cpp:
```diff
@@ -90,7 +90,12 @@ void DumpSigToConsoleBare(MethodContext* mc, CORINFO_SIG_INFO* pSig)
     for (int i = 0; i < (int)pSig->numArgs; i++)
     {
         DWORDLONG          dl;
+#if defined(TARGET_LOONGARCH64)
```
It would be useful to remove all JIT/EE interface changes from this PR so that we can make progress on it independent on https://github.com/dotnet/runtime/pull/62893. Otherwise, the progress on this PR will be blocked by https://github.com/dotnet/runtime/pull/62893.

shushanhf replied:
> It would be useful to remove all JIT/EE interface changes from this PR so that we can make progress on it independent on #62893. Otherwise, the progress on this PR will be blocked by #62893.

If I move the JIT/EE interface to #62893, there will be a lot errors within the CI because they are correlative.

shushanhf replied:
> It would be useful to remove all JIT/EE interface changes from this PR so that we can make progress on it independent on #62893. Otherwise, the progress on this PR will be blocked by #62893.

If I move the JIT/EE interface to #62893, there will be a lot errors within the CI because they are correlative.

jkotas replied:
> If I move the JIT/EE interface to #62893,

Nothing in VM should depend on JIT/EE interface changes. The only part that depends on JIT/EE interface changes should be the JIT. If you exclude JIT/EE interface changes from this PR, the rest of the VM changes should still build fine.

> I want to merge the #62893 into this PR and close the #62893. 

I think it would be better to keep #62893 separate to make it easier to review. If you merge these two PRs together, I expect that the progress will be slower.

shushanhf replied:
OK, I will do it later.

jkotas commented on src/coreclr/vm/callhelpers.cpp:
```diff
@@ -468,11 +468,32 @@ void MethodDescCallSite::CallTargetWorker(const ARG_SLOT *pArguments, ARG_SLOT *
                 {
+#if defined(TARGET_LOONGARCH64)
+                    case 1:
+                        if (m_argIt.GetArgType() == ELEMENT_TYPE_U1)
```
`ELEMENT_TYPE_BOOLEAN` and `ELEMENT_TYPE_CHAR` should be treated as unsigned. They are unsigned types in .NET.

shushanhf replied:
Thanks.

The `ELEMENT_TYPE_CHAR`  is also unsigned ?

jkotas replied:
Yes, `ELEMENT_TYPE_CHAR` is unsigned.

shushanhf replied:
Thanks, I will update it.

jkotas commented on src/coreclr/vm/callhelpers.cpp:
```diff
@@ -468,11 +468,32 @@ void MethodDescCallSite::CallTargetWorker(const ARG_SLOT *pArguments, ARG_SLOT *
+                    case 1:
+                        if (m_argIt.GetArgType() == ELEMENT_TYPE_U1)
+                            *((INT64*)pDest) = (UINT8)pArguments[arg];
+                        else
```
Does this need to handle simple wrappers of primitive types (structs with just one field)?

They should be sign or zero extended based on the underlying type according to this comment: 
https://github.com/loongson/LoongArch-Documentation/pull/32#discussion_r772267496

shushanhf replied:
I think there is no need.
Becasue the field of struct will be pared again while using it.

jkotas replied:
I think you are saying that the comment https://github.com/loongson/LoongArch-Documentation/pull/32#discussion_r772267496 is not valid.

The calling convention details like this have to be very well understood and described. Could you please work on clarifying the ABI for simple wrappers of primitive types with your colleagues?

shushanhf replied:
> I think you are saying that the comment [loongson/LoongArch-Documentation#32 (comment)](https://github.com/loongson/LoongArch-Documentation/pull/32#discussion_r772267496) is not valid.

I am sorry for descriped unclear.
At the comment, using a register passing the `struct {int a;}` is signed extend.

> 
> The calling convention details like this have to be very well understood and described. Could you please work on clarifying the ABI for simple wrappers of primitive types with your colleagues?

This PR is args passing by stack, the struct {...} which sizeof(struct) is less than or equal 4-byte, the upper bits are undefined.
But we signed extend the upper bits except the unsigned types, it will be ok and simply the cases.

jkotas commented:
> What should I do next?

I have added some comments and there are some unaddressed comments from before. It would be great you can address them.

A lot of people on our team who need to review your PRs are taking end-of-year vacation right now, so we won't make much progress on your PRs until after new year.

shushanhf commented:
> > What should I do next?
> 
> I have added some comments and there are some unaddressed comments from before. It would be great you can address them.

Thanks a lot.
I will try my best to do it.

> 
> A lot of people on our team who need to review your PRs are taking end-of-year vacation right now, so we won't make much progress on your PRs until next year.

OK, I will prepare my work at this time.

Merry Christmas,  @jkotas

Commit:
[LoongArch64] add bool type for args' unsigned extention.

shushanhf commented:
In later,
First, I will upload a patch which just update the unsigned extend and rename TARGET_LOONGARCH64 by UNIX
_LOONGARCH64_ABI for ABI.

Then, I will remove the interace  to #62893.

At the last, I will refactor the `ArgDestination and ArgLocDesc` for LoongArch64.

Commit:
[LoongArch64] adding char type for args' unsigned extention.
Also rename `TARGET_LOONGARCH64` to `UNIX_LOONGARCH64_ABI` within ABI.

Commit:
[LoongArch64] amend renaming `TARGET_LOONGARCH64` to `UNIX_LOONGARCH64_ABI` within ABI.

Commit:
[LoongArch64] remove the JIT/EE interface to #62893.

jkotas commented on src/coreclr/vm/callhelpers.cpp:
```diff
@@ -468,11 +468,35 @@ void MethodDescCallSite::CallTargetWorker(const ARG_SLOT *pArguments, ARG_SLOT *
+#if defined(UNIX_LOONGARCH64_ABI)
+                    case 1:
+                    {
+                        CorElementType argType = m_argIt.GetArgType();
```
The size of `ELEMENT_TYPE_CHAR` is 2.

shushanhf replied:
thanks, I will update later.

jkotas commented on src/coreclr/clrdefinitions.cmake:
```diff
@@ -240,6 +240,8 @@ function(set_target_definitions_to_custom_os_and_arch)
       target_compile_definitions(${TARGETDETAILS_TARGET} PRIVATE UNIX_X86_ABI)
     elseif (TARGETDETAILS_ARCH STREQUAL "arm64")
+    elseif (TARGETDETAILS_ARCH STREQUAL "loongarch64")
+      target_compile_definitions(${TARGETDETAILS_TARGET} PRIVATE UNIX_LOONGARCH64_ABI)
```
I do not think that `UNIX_LOONGARCH64_ABI` is useful. There is one and only LoongArch64 ABI so far, so we do not need this define. `TARGET_LOONGARCH64` will work just fine instead.

The `*_ABI` defines are only added as needed for OS-specific ABI differences. For example, there is `OSX_ARM64_ABI` since the OSX ARM64 ABI is different from ARM64 ABI everywhere else (WIndows, Linux, ...). However, there is no `UNIX_ARM64_ABI` ABI since it is not needed.

shushanhf replied:
OK, I will revert it.

shushanhf committed:
[LoongArch64] revert the rename `TARGET_LOONGARCH64` to `UNIX_LOONGARCH64_ABI`.
Also add unsigned extend for CHAR type.

Commit:
[LoongArch64] refactor the `ArgDestination and ArgLocDesc`.

shushanhf commented:
Close & reopen  for trigger CI

shushanhf commented:
@jkotas 
What should I do next for all PRs ?  Which one is first ?

jkotas commented:
All PRs are tagged with area label. The area owners listed in https://github.com/dotnet/runtime/blob/main/docs/area-owners.md are expected to start looking at your PRs.

shushanhf commented:
> All PRs are tagged with area label. The area owners listed in https://github.com/dotnet/runtime/blob/main/docs/area-owners.md are expected to start looking at your PRs.

Thanks for your guidance and help.

shushanhf committed:
Merge branch 'main' into main_loongarch64_2

shushanhf commented:
Hi, @mangod9 
Could you please review this PR and give me some advices ?

shushanhf commented:
The last error is `##[error]The job running on agent Azure Pipelines 12 ran longer than the maximum time of 150 minutes. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134`.

So maybe it's better to test again later.

shushanhf commented:
The last error is `##[error]The job running on agent Azure Pipelines 12 ran longer than the maximum time of 150 minutes. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134`.

So maybe it's better to test again later.

jkotas commented:
> Do you plan on adding some CI validation to ensure builds continue to work for LoongArch64?

@mangod9 Yes, the discussion on this is in https://github.com/dotnet/runtime/issues/59561#issuecomment-991184063 . We are far away from that point. We need to get all the changes reviewed and merged first.

mangod9 commented:
ok sounds good. We should be able to review it this week. Thx!

janvorli commented on src/coreclr/vm/callingconvention.h:
```diff
@@ -789,6 +830,123 @@ class ArgIteratorTemplate : public ARGITERATOR_BASE
+                    assert(!"----------two float;    ? not case ?------------");
+                }
+                else if ((cSlots == 1) && (m_flags & 0x1))
+                    pLoc->m_flag = 1;//first float;
```
Could you please rename the `m_flag` to something more descriptive? It would also be nice to use an enum for the values of m_flag.

The same for `m_flags`.

shushanhf replied:
> Could you please rename the `m_flag` to something more descriptive? It would also be nice to use an enum for the values of m_flag.
> 
> The same for `m_flags`.

Thanks.
How about renaming it by `m_structFloatFlags` as it means for the struct-arg that includes float fields and can be passed by one or two float register(s) ?
For this case, we have to descript the float field's info by this vaues.

shushanhf replied:
> Could you please rename the `m_flag` to something more descriptive? It would also be nice to use an enum for the values of m_flag.
> 
> The same for `m_flags`.

Thanks.
How about renaming it by `m_structFloatFlags` as it means for the struct-arg that includes float fields and can be passed by one or two float register(s) ?
For this case, we have to descript the float field's info by this vaues.

shushanhf replied:
> Could you please rename the `m_flag` to something more descriptive? It would also be nice to use an enum for the values of m_flag.
> 
> The same for `m_flags`.

Thanks.
How about renaming it by `m_structFloatFlags` as it means for the struct-arg that includes float fields and can be passed by one or two float register(s) ?
For this case, we have to descript the float field's info by this vaues.

janvorli replied:
The suggested name sounds good to me.

shushanhf replied:
Thanks

janvorli commented on src/coreclr/vm/callingconvention.h:
```diff
@@ -1522,6 +1684,137 @@ int ArgIteratorTemplate<ARGITERATOR_BASE>::GetNextOffset()
+        }
+        else
+        {
+            MethodTable* methodTablePtr = nullptr;
```
A nit - the convention used for pointers in the runtime is to start the variable with `p`, so the preferred name for this local would be `pMethodTable`.

shushanhf replied:
> A nit - the convention used for pointers in the runtime is to start the variable with `p`, so the preferred name for this local would be `pMethodTable`.

Thanks for your advice and I will rename it.

janvorli commented on src/coreclr/vm/stublink.cpp:
```diff
@@ -1737,6 +1737,9 @@ bool StubLinker::EmitUnwindInfo(Stub* pStubRX, Stub* pStubRW, int globalsize, Lo
             (codeWordsCount << 27);
     } // end else (!m_fProlog)
+
+#elif defined(TARGET_LOONGARCH64)
```
This should not be needed. This code is compiled only for Windows

shushanhf replied:
Thanks.
I will revert it.

janvorli commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -2881,6 +2892,397 @@ void  MethodTable::AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHe
 #endif // defined(UNIX_AMD64_ABI_ITF)
 
+#if defined(TARGET_LOONGARCH64)
+int MethodTable::getFieldTypeByHnd(CORINFO_CLASS_HANDLE clh)
```
This name is misleading. It doesn't get field type, but rather its encoded size and classification. Maybe `getFieldSizeClassificationByHnd` would be a more appropriate name.

shushanhf replied:
Thanks for your guidance~

shushanhf committed:
[LoongArch64] rename the `m_flag` and `getFieldTypeByHnd`.

shushanhf committed:
[LoongArch64] add `NATIVE_SYMBOL_READER_DLL` for compiling error.

shushanhf commented on src/coreclr/vm/argdestination.h:
```diff
@@ -83,6 +83,82 @@ class ArgDestination
+            _ASSERTE(m_argLocDescForStructInRegs->m_cGenReg == 1);
+            _ASSERTE((m_argLocDescForStructInRegs->m_structFields & 3) == 0);//the first field is integer.
+            argOfs = TransitionBlock::GetOffsetOfArgumentRegisters() + m_argLocDescForStructInRegs->m_idxGenReg * 8;
+            *(UINT64*)((char*)m_base + argOfs) = *(UINT64*)src;//NOTE: here ignoring the size.
```
I have a question for help:
when a `struct {int a; float b;}` is passed by registers, why the float is the low 32-bits accessed by `src` rather than the hight 32-bits of `*(UINT64*)src`  ?

Commit:
[LoongArch64] update the version of the `LICENSE description`.

BruceForstall commented:
@shushanhf This PR includes https://github.com/dotnet/runtime/pull/62893; does it need to? Is it up-to-date with that one?

@jkotas @janvorli @mangod9 Who can review this?

Commit:
Merge branch 'main' into main_loongarch64_2

Conflicts:
	src/coreclr/vm/codeman.cpp

Commit:
[LoongArch64] keep same with the comment in `jit/targetloongarch64.h`

BruceForstall commented:
> I think we should wait for #62893 before going back to this one.

@jkotas Now that https://github.com/dotnet/runtime/pull/62893 is merged, who should review this one?

shushanhf commented:
> @shushanhf Can you merge up to current `main`, and fix the conflict?

OK, I will update it now.

Commit:
Merge branch 'main' into main_loongarch64_2

Conflicts:
	src/coreclr/inc/clrnt.h

jkotas commented on src/coreclr/inc/clrnt.h:
```diff
@@ -1030,6 +1030,10 @@ RtlVirtualUnwind(
 
 #endif
 
+#ifndef IMAGE_FILE_MACHINE_LOONGARCH64
```
Number of these `IMAGE_FILE_XXX` definitions are redundant. I am cleaning it up in #65624

shushanhf replied:
> Did you have a chance to think about this?

Yes, some coreclr-test-case runned here. (I have to retest to get which case.)

> Number of these `IMAGE_FILE_XXX` definitions are redundant. I am cleaning it up in #65624

OK, Thanks.
I will update after #65624 .

shushanhf replied:
I update the patch first and I will merge main branch after #65624 merged.

jkotas commented on src/coreclr/vm/exceptionhandling.h:
```diff
@@ -384,6 +384,7 @@ class ExceptionTracker
                     X86_ARG(PT_CONTEXT pContextRecord)
                     ARM_ARG(PT_CONTEXT pContextRecord)
                     ARM64_ARG(PT_CONTEXT pContextRecord)
+                    LOONGARCH64_ARG(PT_CONTEXT pContextRecord)
```
I am deleting this platform-specific arg in #65624

shushanhf replied:
OK, Thanks

jkotas commented:
> @jkotas Now that https://github.com/dotnet/runtime/pull/62893 is merged, who should review this one?

I have submitted my review comments. I would like @janvorli and @mangod9 to take a look as well.

shushanhf commented on src/coreclr/vm/threads.cpp:
```diff
@@ -1136,6 +1136,10 @@ extern "C" void *JIT_WriteBarrier_Table_Loc;
 #endif // TARGET_ARM64
 
+#ifdef TARGET_LOONGARCH64
+extern "C" void *JIT_WriteBarrier_Loc = 0;
```
OK, Thanks

shushanhf commented on src/coreclr/vm/methodtable.h:
```diff
@@ -749,6 +749,10 @@ class MethodTable
     void CheckRunClassInitAsIfConstructingThrowing();
 
+#if defined(TARGET_LOONGARCH64)
+    static int getFieldSizeClassificationByHnd(CORINFO_CLASS_HANDLE clh);
```
Yes, I will rename it.
Thanks

jkotas replied:
The logic in this method seems to be almost exact copy of the JIT/EE interface. That does not sound right.

Could you please structure it similar to the existing getSystemVAmd64PassStructInRegisterDescriptor and related methods?

shushanhf replied:
> The logic in this method seems to be almost exact copy of the JIT/EE interface. That does not sound right.
> 
> Could you please structure it similar to the existing getSystemVAmd64PassStructInRegisterDescriptor and related methods?

Yes, this is also used for getting ABI-info but within the `vm/callingconvention.h` where the JIT/EE interface is unreachable from `vm/callingconvention.h`.
So I have to add new implementation within `methodtable.h`.

shushanhf replied:
Thanks
I will think how to refactor this.

shushanhf replied:
Thanks
I will think how to refactor this.

shushanhf commented on src/coreclr/vm/methodtable.h:
```diff
@@ -1642,6 +1646,9 @@ class MethodTable
 
     DWORD GetIndexForFieldDesc(FieldDesc *pField);
+#if defined(TARGET_LOONGARCH64)
+    CorElementType GetFieldTypeByIndex(DWORD fieldIndex);
```
Yes.
This is used within the `vm/callingconvention.h`.

jkotas replied:
The meaning of index in the implementation of this method is not exactly the same as meaning of index in the existing FieldDesc methods that take or return index..

Would it be better to just use the existing `GetFieldDescByIndex` method in the few places where this is used? ie replace `GetFieldTypeByIndex(...)` with `GetFieldDescByIndex(...).GetFieldType()`?

shushanhf commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -528,6 +528,21 @@ void ExceptionTracker::UpdateNonvolatileRegisters(CONTEXT *pContextRecord, REGDI
+    UPDATEREG(S6);
+    UPDATEREG(S7);
+    UPDATEREG(S8);
+    //UPDATEREG(Tp); //should delete all related!
```
OK, thanks

shushanhf commented on src/coreclr/vm/excep.cpp:
```diff
@@ -6816,6 +6816,10 @@ AdjustContextForJITHelpers(
        // as the faulting address.
        SetIP(pContext, ControlPCPostAdjustment);
+#elif defined(TARGET_LOONGARCH64)
+       //TODO: should confirm for LOONGARCH64.
```
Yes, this had been confirmed by tests and I will move the `TARGET_LOONGARCH64` toghther with ARM64.

shushanhf commented on src/coreclr/vm/gcinfodecoder.cpp:
```diff
@@ -1744,6 +1747,147 @@ OBJECTREF* GcInfoDecoder::GetCapturedRegister(
+{
+    _ASSERTE(regNum >= 1 && regNum <= 31);
+
+    if (regNum == 22)
+    {
+        return (OBJECTREF*) &pRD->pCurrentContext->Fp;
+    }
+    else if (regNum == 1)
+    {
+        return (OBJECTREF*) &pRD->pCurrentContext->Ra;
+    }
```
OK, Thanks

shushanhf commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -4917,6 +4921,62 @@ StackWalkAction SWCB_GetExecutionState(CrawlFrame *pCF, VOID *pData)
                             pES->m_ppvRetAddrPtr = (void **) pRDT->pCallerContextPointers->Lr;
                         }
+#elif defined(TARGET_LOONGARCH64)
+
```
OK, Thanks.

shushanhf committed:
[LoongArch64] amend the code for reviewing.

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1799,6 +1817,46 @@ BOOL Nullable::UnBoxIntoArgNoGC(ArgDestination *argDest, OBJECTREF boxedVal, Met
+                if (destMT == boxedVal->GetMethodTable())
+                {
+                    CopyValueClassArg(argDest, boxedVal->GetData(), destMT, 0);
+                    //if (*(UINT64*)(argDest->GetStructGenRegDestinationAddress()) == 0)
+                    //    *(UINT64*)(argDest->GetStructGenRegDestinationAddress()) = 1;
```
```suggestion
```
Delete this if is not useful

shushanhf replied:
OK, Thanks.
I will delete it next pushing.

Commit:
[LoongArch64] refactor LoongArch64-ABI within `vm`,
and also amend some code for review.

shushanhf committed:
Merge branch 'main' into main_loongarch64_2

shushanhf committed:
[LoongArch64] delete unused codes.

Commit:
Merge branch 'main' into main_loongarch64_2

Conflicts:
	src/coreclr/vm/prestub.cpp

Commit:
[LoongArch64] merge main for #65869.

BruceForstall commented:
> I have submitted my review comments. I would like @janvorli and @mangod9 to take a look as well.

@janvorli @mangod9 Can you please look?

mangod9 commented on src/coreclr/vm/gccover.cpp:
```diff
@@ -76,6 +76,8 @@ bool IsGcCoverageInterruptInstruction(PBYTE instrPtr)
 
 #if defined(TARGET_ARM64)
     instrVal = *reinterpret_cast<UINT32*>(instrPtr);
+#elif defined(TARGET_LOONGARCH64)
```
can this be merged with `TARGET_ARM64` above?

shushanhf replied:
OK, Thanks

am11 commented on src/coreclr/vm/threads.cpp:
```diff
@@ -1136,6 +1136,11 @@ extern "C" void *JIT_WriteBarrier_Table_Loc;
+
 #ifdef TARGET_ARM
 extern "C" void *JIT_WriteBarrier_Loc = 0;
 #endif // TARGET_ARM
```
nit: these if-def blocks (line 1123 to 1146) can be reduced to just:
```c
#ifdef TARGET_X86
extern "C" void *JIT_WriteBarrierEAX_Loc;
#else
extern "C" void *JIT_WriteBarrier_Loc;
void *JIT_WriteBarrier_Loc = 0;
#endif

#ifdef TARGET_ARM64
extern "C" void (*JIT_WriteBarrier_Table)();
extern "C" void *JIT_WriteBarrier_Table_Loc;
void *JIT_WriteBarrier_Table_Loc = 0;
#endif // TARGET_ARM64
```

shushanhf replied:
OK, Thanks

shushanhf committed:
[LoongArch64] amend the format for reviewing.

shushanhf committed:
Merge branch 'main' into main_loongarch64_2

am11 commented on src/coreclr/vm/threads.cpp:
```diff
@@ -1124,22 +1124,15 @@ PCODE AdjustWriteBarrierIP(PCODE controlPc)
 extern "C" void *JIT_WriteBarrierEAX_Loc;
 #else
 extern "C" void *JIT_WriteBarrier_Loc;
+void *JIT_WriteBarrier_Loc = 0;
```
I think this will fix the x64 duplicate symbol error:
```suggestion
#elif TARGET_X64
extern "C" void *JIT_WriteBarrier_Loc;
#else
extern "C" void *JIT_WriteBarrier_Loc;
void *JIT_WriteBarrier_Loc = 0;
```

shushanhf replied:
OK, Thanks.
I will update it now.

shushanhf committed:
[LoongArch64] delete some unused code for reviewing.

shushanhf commented on src/coreclr/vm/frames.h:
```diff
@@ -1834,7 +1845,7 @@ class UnmanagedToManagedFrame : public Frame
     {
         LIMITED_METHOD_DAC_CONTRACT;
-#if defined(TARGET_ARM) || defined(TARGET_ARM64)
+#if defined(TARGET_ARM) || defined(TARGET_ARM64) || defined(TARGET_LOONGARCH64)
```
Thanks.
I will revert this.

BruceForstall commented:
@jkotas @janvorli @mangod9 Do you want to take one more look?

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -4111,10 +4193,11 @@ PTR_RUNTIME_FUNCTION EEJitManager::LazyGetFunctionEntry(EECodeInfo * pCodeInfo)
+            // If we might have fragmented unwind, and we're on ARM64/LoongArch64,
+            // make sure to returning the root record,
+            // as the trailing records don't have prolog unwind codes.
+            pFunctionEntry = FindRootEntry(pFunctionEntry, baseAddress);
```
```suggestion
```
Incorrectly resolved merge conflict?

shushanhf replied:
Thanks.

jkotas commented on src/coreclr/vm/frames.h:
```diff
@@ -860,13 +860,24 @@ class RedirectedThreadFrame : public ResumableFrame
             {
                 fn(ppObj, sc, GC_CALL_INTERIOR | GC_CALL_PINNED);
             }
+#if defined(TARGET_LOONGARCH64)
+            firstIntReg = (Object**)&this->GetContext()->S0;
+            lastIntReg  = (Object**)&this->GetContext()->S8;
+            for (Object** ppObj = firstIntReg; ppObj <= lastIntReg; ppObj++)
+            {
+                fn(ppObj, sc, GC_CALL_INTERIOR | GC_CALL_PINNED);
+            }
+#endif
```
```suggestion
```
It is ok to just include fp in the conservative scanning to keep things simple. Conservative stack scanning is not high performance config.

shushanhf replied:
Ok, thanks.

jkotas commented on src/coreclr/vm/frames.h:
```diff
@@ -860,13 +860,24 @@ class RedirectedThreadFrame : public ResumableFrame
             Object** lastIntReg  = (Object**)&this->GetContext()->X28;
+#elif defined(TARGET_LOONGARCH64)
+            Object** firstIntReg = (Object**)&this->GetContext()->Tp;
+            Object** lastIntReg  = (Object**)&this->GetContext()->X0;
```
```suggestion
            Object** lastIntReg  = (Object**)&this->GetContext()->S8;
```

janvorli commented:
I am planning to take a look. I have a large PR #65738 that will collide with this change, as the precode and VSD stub generating code that's in this change will need to be very different and located in different files. So we'll need to figure out the order in which these two changes go in. It might be better to get in my change first, but there are still some perf details to resolve.

jkotas commented on src/coreclr/vm/gcinfodecoder.cpp:
```diff
@@ -1744,6 +1744,134 @@ OBJECTREF* GcInfoDecoder::GetCapturedRegister(
+
+    return (OBJECTREF*)*(ppReg + regNum);
+#else
+    if(regNum == 1)
+    {
+        return (OBJECTREF*) pRD->pCurrentContextPointers->Ra;
+    }
+    else if(regNum == 22)
+    {
+        return (OBJECTREF*) pRD->pCurrentContextPointers->Fp;
+    }
+    else if (regNum < 22)
+    {
+        return (OBJECTREF*)*(DWORD64**)(&pRD->volatileCurrContextPointers.A0 + (regNum - 4));//A0=4.
+    }
+
+    return (OBJECTREF*)*(DWORD64**)(&pRD->pCurrentContextPointers->S0 + (regNum-23));
```
```suggestion
    if (regNum >= 4 && regNum < 22)
    {
        return (OBJECTREF*)*(DWORD64**)(&pRD->volatileCurrContextPointers.A0 + (regNum - 4));//A0=4.
    }

    return (OBJECTREF*)*(DWORD64**)(&pRD->pCurrentContextPointers->R0 + regNum));
```
Can this be simplified like this?

shushanhf replied:
Thanks

jkotas commented on src/coreclr/vm/gcinfodecoder.cpp:
```diff
@@ -1744,6 +1744,134 @@ OBJECTREF* GcInfoDecoder::GetCapturedRegister(
+{
+    GCINFODECODER_CONTRACT;
+
+    //_ASSERTE(regNum != 1);
```
```suggestion
```

janvorli commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -2881,6 +2881,390 @@ void  MethodTable::AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHe
 #endif // defined(UNIX_AMD64_ABI_ITF)
 
+#if defined(TARGET_LOONGARCH64)
+int MethodTable::getLoongArch64PassStructInRegisterFlags(CORINFO_CLASS_HANDLE cls)
```
A nit - can you please start the method name with an upper case letter to match the coding convention?

shushanhf replied:
OK, Thanks

shushanhf commented on src/coreclr/vm/gccover.cpp:
```diff
@@ -114,6 +114,11 @@ bool IsOriginalInstruction(PBYTE instrPtr, GCCoverageInfo* gcCover, DWORD offset
         return (instrVal == origInstrVal);
     }
+#elif defined(TARGET_LOONGARCH64)
+    _ASSERTE(!"not implemented for LOONGARCH64 platform");
```
I will delete it for old code and will merge above with arm64.
Thanks

shushanhf committed:
[LoongArch64] amend code for CR feedback @jkotas @janvorli

shushanhf commented on src/coreclr/vm/callingconvention.h:
```diff
@@ -788,6 +829,107 @@ class ArgIteratorTemplate : public ARGITERATOR_BASE
+
+            if ((GetArgType() == ELEMENT_TYPE_R4) || ((m_flags == STRUCT_FLOAT_FIELD_ONLY_ONE) && (GetArgSize() == 4)))
+            {
+                pLoc->m_structFloatFlag = 5;
```
I will update this later.
Thanks

shushanhf commented on src/coreclr/vm/callingconvention.h:
```diff
@@ -788,6 +829,107 @@ class ArgIteratorTemplate : public ARGITERATOR_BASE
+        return;
+    }
+
+    int             m_idxGenReg;        // Next general register to be assigned a value
```
OK, Thanks

Commit:
[LoongArch64] add class type for LoongArch64-ABI.

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -2881,6 +2881,523 @@ void  MethodTable::AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHe
+                else if (fieldType == ELEMENT_TYPE_CLASS)
+                {
+                    MethodTable* pMethodTable2  = pFieldStart->GetFieldTypeHandleThrowing().GetMethodTable();
+                    if (pMethodTable2->IsArray())
```
It does not make sense to special case arrays for the field layout. Why did you add special casing arrays here? 

Note that arrays in .NET are regular objects. They are not inline like in C/C++.

shushanhf replied:
If always within managed code there is no need to add this, for calling unmanaged code while using LoongArch64's ABI, 
liking
```
[StructLayout(LayoutKind.Sequential)]
public unsafe struct FixedArrayClassificationTest
{
    [MarshalAs(UnmanagedType.ByValArray, SizeConst = 3)] 
    public Int32Wrapper[] arr;
    public float f;
}

[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
public struct UnicodeCharArrayClassification
{
    [MarshalAs(UnmanagedType.ByValArray, SizeConst = 6)] 
    public char[] arr;
    public float f;
}
```
which the field of struct is array type, LoongArch64's ABI has to parse the field's type.

jkotas replied:
Right, but the code here does not deal with the unmanaged view of the type. It deals with the managed view of the type where the array is regular object reference.

shushanhf replied:
After refactor, this interface will be called also by managed from the jitinterface.
It will be called within the file `vm/callingconvention.h`.

jkotas replied:
Right, but that does not explain the special casing of arrays in this spot. Do you have a test case that fails when you delete the special casing of the arrays here?

shushanhf replied:
> Right, but that does not explain the special casing of arrays in this spot.


> Do you have a test case that fails when you delete the special casing of the arrays here?

The following code is from the case `baseservices/TieredCompilation/BasicTestWithMcj/BasicTestWithMcj.sh`

> ```
> [StructLayout(LayoutKind.Sequential)]
> public unsafe struct FixedArrayClassificationTest
> {
>     [MarshalAs(UnmanagedType.ByValArray, SizeConst = 3)] 
>     public Int32Wrapper[] arr;
>     public float f;
> }
> 
> [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
> public struct UnicodeCharArrayClassification
> {
>     [MarshalAs(UnmanagedType.ByValArray, SizeConst = 6)] 
>     public char[] arr;
>     public float f;
> }
> ```

shushanhf replied:
@jkotas 
I update the previous comment about the test-case name, `src/tests/Interop/StructMarshalling/PInvoke/MarshalStructAsLayoutSeq.cs`

shushanhf replied:
I will update this PR later for amending the LA-ABI after merged the implementation of `CEEInfo::getLoongArch64PassStructInRegisterFlags` and `MethodTable::GetLoongArch64PassStructInRegisterFlags` which I had tested on LoongArch64-linux by the SDK6.0.102 with the same test-cases passed.

jkotas replied:
The special casing of arrays should be only done when `IsNativeValueType() == true`. It should be enough to make this test case work.

This path is for `IsNativeValueType() == false`. It should not have any special casing for arrays.

shushanhf replied:
@jkotas

shushanhf replied:
@jkotas 
Thanks very much.
I will amend this.

shushanhf replied:
Yesterday I just only runned all the release test cases as release is faster and all test had the same passed.
Today there are two assert error within the debug, so I update it again after amend to fix them.

shushanhf replied:
I had finished the tests on my local machine to test the LoongArch64.

Please give me some advices if you review further.
Thanks

shushanhf committed:
[LoongArch64] Amend the LoongArch64's ABI
after merged `MethodTable::GetLoongArch64PassStructInRegisterFlags()`
and `CEEInfo::getLoongArch64PassStructInRegisterFlags()`

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -2881,6 +2881,515 @@ void  MethodTable::AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHe
+                else if ((fieldType == ELEMENT_TYPE_CLASS) && pMethodTable->IsArray())
+                {
+                    MethodTable* pMethodTable2  = pFieldStart->GetFieldTypeHandleThrowing().GetMethodTable();
+                    if (pMethodTable2->IsArray())
+                        size = GetLoongArch64PassStructInRegisterFlags((CORINFO_CLASS_HANDLE)pMethodTable2);
```
Is this needed? I believe that `GetLoongArch64PassStructInRegisterFlags` is alway going to return `STRUCT_NO_FLOAT_FIELD` for arrays.

shushanhf replied:
You are right.
I had deleted it within the latest PR.
Thanks

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -2881,6 +2881,515 @@ void  MethodTable::AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHe
+                    }
+                    if (pMethodTable2->IsArray())
+                    {
+                        if ((pMethodTable2->GetNumInstanceFieldBytes() / pFieldStart[0].GetSize())  > 1)
```
`GetNumInstanceFieldBytes` for arrays should be always zero. Is this path ever hit?

shushanhf replied:
> `GetNumInstanceFieldBytes` for arrays should be always zero. Is this path ever hit?

I remembered that the array cases hitted it. I use this is to get the count of the element of array.
I will check it now.

shushanhf replied:
> > `GetNumInstanceFieldBytes` for arrays should be always zero. Is this path ever hit?
> 
> I remembered that the array cases hitted it. I use this is to get the count of the element of array. I will check it now.

I have checked again liking
```
(gdb) p *pFieldStart
$2 = {m_pMTOfEnclosingClass = {static isRelative = false, m_ptr = 1097319343376}, m_mb = 151, m_isStatic = 0, m_isThreadLocal = 0, m_isRVA = 0, m_prot = 6,
  m_requiresFullMbValue = 0, m_dwOffset = 0, m_type = 18, m_debugName = 0xfff4116f18 "arr"}
(gdb) p *pMethodTable2    
$5 = {static MaxGenericParametersForSpecialMarkerType = 8, static s_pMethodDataCache = 0xaaaab93f30, static s_fUseParentMethodData = 1, static s_fUseMethodDataCache = 1,
  static UNION_MASK = 3,
  static c_DispatchMapSlotOffsets = 0xfff77d5f9e <MethodTable::c_DispatchMapSlotOffsets> "8@8H8@8H@HHP8@8H@HHP@HHPHPPXHHHHHHHPHHHPHPPXHHHPHPPXHPPXPXX`N11MethodTable16MethodDataObjectE", static c_NonVirtualSlotsOffsets = <same as static member of an already seen type>,
  static c_ModuleOverrideOffsets = <same as static member of an already seen type>,
  static c_OptionalMembersStartOffsets = <same as static member of an already seen type>, m_dwFlags = 2148139012, m_BaseSize = 24, m_wFlags2 = 11, m_wToken = 0,
  m_wNumVirtuals = 24, m_wNumInterfaces = 6, debug_m_szClassName = 0xff7d549018 "Int32Wrapper[]", m_pParentMethodTable = {static isRelative = false,
    m_addr = 1097312007520}, m_pLoaderModule = {static isRelative = false, m_ptr = 1097316892360}, m_pWriteableData = {static isRelative = false, m_ptr = 1097319354336},
  {m_pEEClass = {static isRelative = false, m_ptr = 1097319353744}, m_pCanonMT = {static isRelative = false, m_ptr = 1097319353744}}, {m_pPerInstInfo = {
      static isRelative = false, m_ptr = 1097319345880}, m_ElementTypeHnd = 1097319345880, m_pMultipurposeSlot1 = 1097319345880}, {m_pInterfaceMap = {
      static isRelative = false, m_ptr = 1097319354032}, m_pMultipurposeSlot2 = 1097319354032}}
(gdb) p ((MethodTable*)0xff7d548e40)->IsArray()
$7 = 1                        
(gdb) p pMethodTable2->GetNumInstanceFieldBytes()
$6 = 24                       
(gdb)
```

The above test case is `src/tests/Interop/StructMarshalling/PInvoke/MarshalStructAsLayoutSeq.cs` and the struct is https://github.com/dotnet/runtime/pull/62885#discussion_r819261350

jkotas replied:
`GetNumInstanceFieldBytes() == 24` is the size of object header + methodtable + array size field. It is not related in any way to how many elements are in the array. If this condition makes `MarshalStructAsLayoutSeq` test pass, it is just a happy accident. It is going to fail if the actual size of the array is different.

This condition is not right.

shushanhf replied:
> `GetNumInstanceFieldBytes() == 24` is the size of object header + methodtable + array size field. It is not related in any way to how many elements are in the array. If this condition makes `MarshalStructAsLayoutSeq` test pass, it is just a happy accident. It is going to fail if the actual size of the array is different.
> 
> This condition is not right.

Thanks very much.
I will modify it.

shushanhf replied:
@jkotas 
Thanks very much for your review and guidance.

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -2881,6 +2881,515 @@ void  MethodTable::AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHe
+
+                    if (pMethodTable2->IsArray())
+                    {
+                        if ((pMethodTable2->GetNumInstanceFieldBytes() / pFieldStart[1].GetSize())  > 1)
```
Same here

shushanhf replied:
Same with https://github.com/dotnet/runtime/pull/62885#discussion_r821249868

BruceForstall commented:
@shushanhf Now that https://github.com/dotnet/runtime/pull/65738 has merged, this needs to be updated.

shushanhf commented:
> @shushanhf Now that #65738 has merged, this needs to be updated.

OK, Thanks

shushanhf committed:
Merge branch 'main' into main_loongarch64_2
For porting #65738 to LoongArch64.

BruceForstall commented:
@jkotas @janvorli Time for another (final?) code review here?

shushanhf committed:
[LoongArch64] Fix the assert error
when running hello-world within debug-mode after refacting.

janvorli commented on src/coreclr/vm/loongarch64/thunktemplates.S:
```diff
@@ -0,0 +1,38 @@
+
+LEAF_ENTRY FixupPrecodeCode
+    pcaddi  $r21, 0x1003  //3, for Type encoding.
+    ld.d  $r21,$r21, (FixupPrecodeData__Target - 4*3)
```
Just to make sure I understand this code - is the page size of the processor 16kB?

shushanhf replied:
Yes, you are right.

I had tested this PR based on the lastest runtime yesterday and it is ok which based on your refactor.

shushanhf commented:
Now I had tested the hello-world can passed both the release and debug mode based on the latest main-LoongArch64.

BruceForstall commented:
@mangod9 @janvorli @jkotas Is there anything left here to review? (fyi, I just merged the LoongArch64 JIT PR)

shushanhf commented:
@mangod9 @janvorli @jkotas
Could you please give me some advices further?

shushanhf commented:
Hi, @mangod9 @janvorli @jkotas

What should I dofor this PR？
Could you please give me some advices ?

mangod9 commented:
Hi @shushanhf, sorry for delayed response. We plan to review this again this week and will merge with a green CI. Thanks

janvorli reviewed (approved):
LGTM, thank you!

mangod9 commented:
Some JIT HW Intrinsics tests are failing -- I will submit a rerun. 
cc @BruceForstall in case these have been seen before: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-62885-merge-13bbab73369a44af9a/JIT.HardwareIntrinsics/1/console.c2df8b25.log?sv=2019-07-07&se=2022-04-28T05%3A53%3A25Z&sr=c&sp=rl&sig=B24gI9K1nVgOTBqIiWnGKoSByMPjTbvq0VZSpMHTCbc%3D

mangod9 commented:
Looks to have passed on rerun. So will merge it now.

BruceForstall commented:
@shushanhf Thanks for your patience and response to code review feedback. And congratulations on getting this merged!
